{:title
 "JUnit Tip: Verifying that an Exception with a Particular Message was Thrown",
 :date "2011-09-16",
 :layout :post,
 :tags ["testing"],
 :tags-orig ["junit"],
 :categories ["Testing"],
 :extra-css #{}
}

++++
JUnit has a hidden treasure which makes it easy to do something we have long longed for - namely not only to verify that an exception of a particular type has been thrown but also that its message contains the expected message. The hidden pearl is the <a href="https://kentbeck.github.com/junit/javadoc/latest/index.html?org/junit/rules/ExpectedException.html">@Rule ExpectedException</a> and its JavaDoc documents well how to use it (slightly modified):<br><br><!--more--><br><br><pre><code>
import org.junit.*;
import org.junit.rules.ExpectedException;<br><br>public static class HasExpectedException {
        @Rule
        public ExpectedException thrown= ExpectedException.none();<br><br>        @Test
        public void throwsNothing() {
            // no exception expected, none thrown: passes.
        }<br><br>        @Test
        public void throwsNullPointerExceptionWithMessage() {
                thrown.expect(NullPointerException.class);
                thrown.expectMessage(&quot;What happened here?&quot;);
                thrown.expectMessage(allOf(containsString(&quot;What&quot;), containsString(&quot;here&quot;)));
                throw new NullPointerException(&quot;What happened here?&quot;);
        }
 }
</code></pre><br><br>(As you might have noticed, it uses <a href="https://code.google.com/p/hamcrest/wiki/Tutorial">Hamcrest matchers</a>; containsString isn't included directly in junit and thus you'd need junit-dep + hamcrest jars.)
++++
